### PR TITLE
DQN

### PR DESCRIPTION
**Pull request by Het-Shah**
Used MlpValue as the model.

Refer #36

### CODE REVIEW

**Comment by Het-Shah** [comment]
Travis Build fails because of PPO i guess. Will have to update PPO first

File: jigglypuffRL/deeprl/dqn/dqn.py [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+        env,
+        double_dqn=False,
+        epochs=100,
+        max_iterations_per_epoch=100,
+        max_ep_len=1000,
+        gamma=0.99,
+        lr=0.001,
+        batch_size=32,
+        replay_size=100,
+        plot_loss_reward_graph=False,
```
Comment (ajaysub110) [inline jigglypuffRL/deeprl/dqn/dqn.py:50]: Loss and reward plots will be on tensorboard anyways right? Is this required then?

File: jigglypuffRL/deeprl/dqn/dqn.py [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+            self.model = MlpValue(
+                self.env.observation_space.shape[0],
+                self.env.action_space.n,
+                val_type="Qs",
+            )
+            self.target_model = MlpValue(
+                self.env.observation_space.shape[0],
+                self.env.action_space.n,
+                val_type="Qs",
+            )
```
Comment (ajaysub110) [inline jigglypuffRL/deeprl/dqn/dqn.py:116]: The implementation of the model looks slightly incorrect. The regular DQN uses 2 networks, `Q` an `targ_Q`. Double DQN uses 2 independent Q function estimates which means that it has 4 networks `Q, targ_Q, Q', targ_Q'`

File: jigglypuffRL/deeprl/dqn/dqn.py [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+
+        if self.double_dqn and (network_type == "MLP"):
+            self.model = MlpValue(
+                self.env.observation_space.shape[0],
+                self.env.action_space.n,
+                val_type="Qs",
+            )
```
Comment (ajaysub110) [inline jigglypuffRL/deeprl/dqn/dqn.py:111]: To avoid code duplication and account for all cases, you could do something like 
```python
if network_type == 'MLP':
  self.q = MlpValue(...)
  self.q_t = deepcopy(Q)

  if self.double_dqn:
    self.q2 = MlpValue(...)
    self.q2_t = deepcopy(...)
```
This is still pretty tedious, but as of now it should be fine

File: jigglypuffRL/deeprl/dqn/dqn.py [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+
+        q_value = q_values.gather(1, action.unsqueeze(1)).squeeze(1)
+        next_q_value = next_q_values.max(1)[0]
+        expected_q_value = reward + self.gamma * next_q_value * (1 - done)
+
+        loss = F.smooth_l1_loss(q_value, expected_q_value)
+
+        self.loss_hist.append(loss)
+
+        return loss
```
Comment (ajaysub110) [inline jigglypuffRL/deeprl/dqn/dqn.py:161]: This too needs be corrected since the Double DQN thing is wrong

File: jigglypuffRL/deeprl/dqn/dqn.py [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+        plt.figure(figsize=(20, 5))
+        plt.subplot(131)
+        plt.title(
+            "Average reward for last 10 frames: %s" % (np.mean(self.reward_hist[-10:]))
+        )
+        plt.plot(self.reward_hist)
+        plt.subplot(132)
+        plt.title("loss")
+        plt.plot(self.loss_hist)
+        plt.show()
```
Comment (ajaysub110) [inline jigglypuffRL/deeprl/dqn/dqn.py:184]: Not required since we're already using tensorboard

File: jigglypuffRL/deeprl/dqn/dqn.py [hunk 6]
```diff
@@ -0,0 +1,239 @@
...
+                self.update_target_model()
+
+        if self.plot_loss_reward_graph == True:
+            self.plot()
+
+        if self.tensorboard_log:
+            self.writer.close()
+
+        if self.render:
+            self.env.close()
```
Comment (ajaysub110) [inline jigglypuffRL/deeprl/dqn/dqn.py:233]: `self.env.close()` needs to be done regardless of whether `render` was `True` or `False`

File: jigglypuffRL/deeprl/dqn/dqn.py (hunk 5 above)
Comment (Het-Shah) [reply jigglypuffRL/deeprl/dqn/dqn.py:184]: Cool will remove this was not using tensorboard, this was my sort of helper function.

File: jigglypuffRL/deeprl/dqn/dqn.py (hunk 3 above)
Comment (Het-Shah) [reply jigglypuffRL/deeprl/dqn/dqn.py:111]: There will be an exponential in increase the number of models with more variants I suppose. I will account for all the cases using `self.model` and `self.target_model` probably.

Comment (ajaysub110) [reply jigglypuffRL/deeprl/dqn/dqn.py:111]: Don't worry, I've fixed in my next PR. Will be pushing now. Use that preferably

Comment (ajaysub110) [reply jigglypuffRL/deeprl/dqn/dqn.py:111]: I've added a get_model function which can be called directly from the algo class. It'll make things much more convenient hopefully

Comment (ajaysub110) [reply jigglypuffRL/deeprl/dqn/dqn.py:111]: See #57

File: jigglypuffRL/deeprl/dqn/dqn.py (hunk 2 above)
Comment (Het-Shah) [reply jigglypuffRL/deeprl/dqn/dqn.py:116]: In double DQN I don't think we will need 4 networks. According to this paper

https://arxiv.org/pdf/1509.06461.pdf

**Comment by ajaysub110** [comment]
Oh okay. Looks like they don't use target networks here. Then only 2
networks should be required right?
So in this case, for regular dqn, there should be a Q and targQ. For double
case, we need Q and Q2.

**Comment by Het-Shah** [comment]
I will resolve these merge conflicts. Have a look at the loss functions now. This makes sense to me according to the paper I mentioned above.

File: jigglypuffRL/deeprl/dqn/dqn.py (hunk 3 above)
Comment (Het-Shah) [reply jigglypuffRL/deeprl/dqn/dqn.py]: will take a look on this

**Comment by Het-Shah** [comment]
someone review this lol

**Comment by Het-Shah** [comment]
Added Dueling DQN

**Comment by Het-Shah** [comment]
will remove these LGTM alerts with next commit

File: jigglypuffRL/deeprl/dqn/dqn.py [hunk 7]
```diff
@@ -0,0 +1,225 @@
...
+    def create_model(self, network_type):
+        if network_type == "MLP":
+            self.model = get_model("v", network_type)(
+                self.env.observation_space.shape[0], self.env.action_space.n, "Qs"
+            )
+            self.target_model = deepcopy(self.model)
+            
+            if self.dueling_dqn:
+                self.model = DuelingDQNValueMlp(self.env.observation_space.shape[0], self.env.action_space.n)
+                self.target_model = deepcopy(self.model)
```
Comment (Sharad24) [inline jigglypuffRL/deeprl/dqn/dqn.py:109]: Is `double_q` missing here?

File: jigglypuffRL/deeprl/dqn/dqn.py (hunk 2 above)
Comment (Sharad24) [reply jigglypuffRL/deeprl/dqn/dqn.py]: @ajaysub110 I think we just need 2 networks for both the cases. If you know if any open source implementation with 3/4 networks please give us a pointer to it.

**Comment by Sharad24** [comment]
Btw, Can you add a `get_hyperparams` function to return the important hyperparams?

File: jigglypuffRL/deeprl/dqn/dqn.py (hunk 7 above)
Comment (Het-Shah) [reply jigglypuffRL/deeprl/dqn/dqn.py:109]: Nope it is same for dqn and double_dqn

**Comment by Het-Shah** [comment]
> Btw, Can you add a `get_hyperparams` function to return the important hyperparams?

I will add this at the end once all the variants are done.

**Comment by Sharad24** [comment]
Right, I missed the target model definition.
